### PR TITLE
Modify some classes to functions and struct in DEM

### PR DESCRIPTION
**Pull request by acdaigneault**
### Description

Some classes should not has been classes but functions or struct.
I modified those classes to a more coherent architechture.
I also gathered all the particle-object contact info structures in one file because it is more convenient that having 3 files for small structs.

### Testing

Unit tests have been change according to that.

### Miscellaneous (will be removed when merged)

I do not think that needs a change log.
The particle-wall contact info needs to be changed as other, but I will do it when modifying the particle-wall forces.

### Checklist (will be removed when merged)
See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#pull-requests) for more information about the pull request process.

Code related list:
- [x] All in-code documentation related to this PR is up to date (Doxygen format)
- [x] Lethe documentation is up to date
- [x] The branch is rebased onto master
- [x] Code is indented with indent-all and .prm files (examples and tests) with prm-indent
- [x] If parameters are modified, the tests and the documentation of examples are up to date
- [x] Changelog (CHANGELOG.md) is up to date if the refactor affects the user experience or the codebase

Pull request related list:
- [x] No other PR is open related to this refactoring
- [x] Labels are applied
- [x] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [x] If this PR closes an issue or is related to a project, it is linked in the "Projects" or "Development" section
- [x] If any future works is planed, an issue is opened
- [x] The PR description is cleaned and ready for merge

### CODE REVIEW

File: source/dem/CMakeLists.txt [hunk 1]
```diff
@@ -50,6 +50,7 @@ add_library(lethe-dem
   # Headers
   ../../include/dem/adaptive_sparse_contacts.h
   ../../include/dem/boundary_cells_info_struct.h
+  ../../include/dem/contact_info.h
```
Comment (acdaigneault) [inline source/dem/CMakeLists.txt:53]: I made it! Alphabetical order!

File: include/dem/particle_particle_broad_search.h [hunk 2]
```diff
@@ -18,164 +18,121 @@
...
-          contact_pair_candidates.emplace(main_particle_id,
-                                          candidates_container);
-        candidates_container_it = pair_it_bool.first;
-      }
+/**
+ * @brief Finds a vector of pairs (particle_particle_candidates) which contains the
+ * candidate particle-particle collision pairs. These collision pairs will be
+ * used in the fine search to investigate if they are in contact or not.
+ * This version of the function is used when adaptive sparse contacts regards
+ * mobility is enable.
```
Comment (blaisb) [inline include/dem/particle_particle_broad_search.h:95]: what you mean regards mobility?

**Review by blaisb — COMMENTED** [review]
Some minor comments. I will continue reading tomorrow but  I doubt I will have many comments

File: include/dem/dem_contact_manager.h [hunk 3]
```diff
@@ -299,20 +300,6 @@ class DEMContactManager
 
   // Containers with other information
   typename DEM::dem_data_structures<dim>::cell_vector periodic_cells_container;
-
-private:
-  // Broad search objects
-  ParticleParticleBroadSearch<dim>  particle_particle_broad_search_object;
```
Comment (blaisb) [inline include/dem/dem_contact_manager.h:305]: Cool to get rid of all these objects :)

File: include/dem/find_cell_neighbors.h [hunk 4]
```diff
@@ -26,139 +26,124 @@
...
-      &cells_ghost_periodic_neighbor_list,
-    typename DEM::dem_data_structures<dim>::cells_neighbor_list
-      &cells_ghost_local_periodic_neighbor_list);
+/**
+ * @brief Finds the periodic neighbor list (without repetition) of all the
+ * active cells in the triangulation. It gets the coinciding vertices of the
+ * cells at the periodic boundary 0 to get lists of the periodic neighbor
+ * cells on the periodic boundary 1 for each cells. There is some check to
+ * prevent repetition of a cell in a list (up to 4 vertices can have the
+ * same cell in common in 3D). Also, 3 types of container are used for
```
Comment (blaisb) [inline include/dem/find_cell_neighbors.h:59]: isnt that 8 also?

**Review by blaisb — APPROVED** [review]
last few comments. This is very good work. Thanks for this amazing refactor. It's much easier to read now.

**Comment by blaisb** [comment]
Some tests seem to be crashing but it seems to be due to a deal.II update. I'll try to work on this tomorrow AM

File: include/dem/find_cell_neighbors.h [hunk 5]
```diff
@@ -26,139 +26,124 @@
...
- * find_cell_neighbors() is written to avoid any repetition, for instance if
- * cell B is recognized as the neighbor of cell A once, cell A will not appear
- * in the neighbor list of cell B again. On the other hand,
- * find_full_cell_neighbors() function finds the neighbors list with this
- * repetition.
+ * @brief Finds the neighbor list (without repetition) of all the active
+ * cells in the triangulation. It gets the vertices of the cells to get lists
+ * of the neighbor for each cells. There is some check to prevent repetition
+ * of a cell in a list (up to 4 vertices can have the same cell in common in
+ * 3D). 2 types of container are used for cell neighbors : local-local cells
```
[the remark this answers is not in the document]
Comment (acdaigneault) [reply include/dem/find_cell_neighbors.h:33]: Ok! I did not modify the text here, but I can take a look so it makes more sense

File: include/dem/dem_contact_manager.h (hunk 3 above)
Comment (acdaigneault) [reply include/dem/dem_contact_manager.h:305]: So many!

File: include/dem/find_cell_neighbors.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (acdaigneault) [reply include/dem/find_cell_neighbors.h]: I agree the description is not that obvious, but this is still a good summary. The comments in the function itself are quite explicit tho

**Comment by blaisb** [comment]
@acdaigneault can you rebase, this way we will be able to see what is breaking because of previous bugs and not

**Comment by blaisb** [comment]
@acdaigneault I'm not sure what is this hang error you are getting, but it seems something is off since it happens in both release and debug?

**Comment by acdaigneault** [comment]
> @acdaigneault I'm not sure what is this hang error you are getting, but it seems something is off since it happens in both release and debug?

Yes this is related to my PR

**Review by voferreira — APPROVED** [review]
All clear to me. Cleaning is so satisfying haha. Great job!

**Pull request merged**